Update atomic_rmw operations to use acq_rel ordering; introduce fence

eschnett opened the pull request:
- set memory ordering for atomic loads to acquire
- set memory ordering for atomic stores to release
- set memory ordering for atomic rmw operations to acq_rel
- introduce fence instruction with acq_rel ordering

See #15106 for a related discussion.

yuyichao commented on test/threads.jl:
```diff
@@ -141,3 +141,40 @@ end
+    end
+end
+function test_fence()
+    commbuf = CommBuf()
```
Initialize `data1` and `data2` ?

eschnett replied:
Atomics are automatically initialized with zero. I'll add a manual initialization anyway.

yuyichao replied:
`.data1` and `.data2` aren't initialized though. (they are `#undef` by default).

yuyichao replied:
And the `.correct` field is randomly initiaized.

eschnett replied:
Okay, at least `.correct` was initialized in a different routine.

I re-wrote the test, and everything should be properly initialized now.

yuyichao commented on test/threads.jl:
```diff
@@ -141,3 +141,40 @@ end
+    for i in 10000:-1:1
+        commbuf.data1[] = i
+        fence()
+        commbuf.data2[] = i
```
Does this fail without the fence? IIUC, since the store to `data2` has a release order, the store to `data1` in the same loop has to be visible when the store to `data2` is visible.

eschnett replied:
This fence is a no-op on x86 -- so no, I couldn't observe failure yet. We'll have to test on a different architecture.

yuyichao commented on base/atomics.jl:
```diff
@@ -68,8 +69,13 @@ for (typ, lt) in atomicintsmap
         end
         @eval $fn(x::Atomic{$typ}, v::$typ) =
             llvmcall($"""
-                    %rv = atomicrmw volatile $rmw $lt* %0, $lt %1 acquire
```
Any reason to leave the `volatile` here but not other ones?

eschnett replied:
No reason. I removed it.

kpamnany commented on base/atomics.jl:
```diff
@@ -10,7 +10,8 @@ export
     atomic_and!, atomic_nand!, atomic_or!, atomic_xor!,
-    atomic_max!, atomic_min!
+    atomic_max!, atomic_min!,
+    fence
```
Call this ``atomic_fence`` for consistency?

eschnett replied:
Done.

kpamnany commented:
Thanks @eschnett for raising this PR.

yuyichao commented on test/threads.jl:
```diff
@@ -141,3 +141,41 @@ end
+    CommBuf() = new(Atomic{Int}(0), Vector{Float64}(100), false)
+end
+function test_fence_write(commbuf::CommBuf)
+    sleep(0.1)
```
This is danguous since we don't really support task switch on a thread. We happens to run `i == 1` on the master thread but this should probably `ccall` libc `sleep` instead. (I believe the `@threadcall` test has a version that work on windows too).

eschnett replied:
I simply want to ensure that the other thread -- which depends on this one -- actually needs to wait. We can use a condition variable instead, or some other delay.

yuyichao replied:
Conditional variable triggered by the other thread will not work since we don't garentee the other iteration will be running concurrently with this one (especially if threading is disabled and probably not even with threading enabled in the future).

eschnett replied:
I think the safest is to just ensure a small delay. I'm thinking of busy-waiting via `time()`.

yuyichao replied:
Why not `ccall`ing `sleep`?

eschnett replied:
Sure -- if this works. It just sounded complicated.

yuyichao replied:
what's complicated about it? It's literally two lines of code (one for unix one for windows) each with a single ccall and no loops. https://github.com/JuliaLang/julia/blob/7b1cba7fb85e590666cce17dcaa85971455b5b9b/test/ccall.jl#L507

yuyichao commented on test/threads.jl:
```diff
@@ -141,3 +141,41 @@ end
+        commbuf.data[i] = i
+    end
+    atomic_fence()
+    commbuf.barrier[] = 1
```
To actually test the fence, I think you want non atomic store here. Otherwise the acquire/release alone should garentee the order in the two threads.

This test (with the `atomic_fence()` removed) should be a good test for the store and load schematics though.

eschnett replied:
Atomic fences on different threads are only ordered with respect to each other if there are ordered atomic writes/reads in between. The fence then orders the preceding non-atomic writes (on one thread) and the following non-atomic reads (on the other thread).

Unfortunately, on x86, the fence is a no-op, since the hardware always makes this guarantee anyway.

yuyichao replied:
I don't think that's right.

Observing the result of an atomic release store with an atomic acquire load grantees that any memory operations on the thread performing the release store that happens before the store in source order can be observed on the thread performing the acquire load after the load in source order so in this case the write to the data will be available to the read thread as long as the release store to the barrier is after the write to the data (in source order) and as long as the read of data is after the acquire load. See the last code example in the acquire/release order in this GCC wiki https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync .

yuyichao replied:
P.S. I think the following sentence on the GCC wiki is confusing,

> Since 'y' is not an atomic variable, the store to 'y' happens-before the store to 'x'

I thnk the only thing that matters is that the two happens on the same thread and the second one is an atomic release store. Whether `y` is atomic or not doesn't matter here (and making it atomic shouldn't make the assertion fail.....).

eschnett replied:
Yes, my fence test isn't working. Thanks for the pointer. To make it work, I need two non-atomic writes separated by a fence in one thread, observed by another thread.

yuyichao replied:
> I need two non-atomic writes separated by a fence in one thread, observed by another thread.

I think that's correct. As I mentioned above, I think you can keep this test (without the fence) to test the memory order of the atomic store and load.

eschnett replied:
I can't think of a way to write a fence test.

Fences are only interesting for stores or loads that are not already ordered. Our atomic operators are currently all ordered. On the other hand, non-atomic loads and stores cannot be directly observed from another thread (as this is undefined), and ensuring the proper ordering requires ordered atomic load/stores, which then remove the need for a fence.

I think acquire-release fences only make sense if there are unordered atomic operations. We currently don't provide these in Julia.

I am thus changing the fence to a sequential-consistency fence.

yuyichao replied:
I think when you are dealing with only two threads that both have fences, `acq_rel` and `seq_cst` doesn't make any difference so a `acq_rel` fence should be enough. I can't be 100% sure here though so @ArchRobison @kpamnany

yuyichao commented on test/threads.jl:
```diff
@@ -141,3 +141,91 @@ end
+
+# Test atomic memory ordering
+type CommBuf{T}
+    var1::Ref{T}
```
Why is `Ref` necessary? This is already a mutable type.

eschnett replied:
I want to ensure that the atomic variables live in different cache lines, thus I allocate them as arrays with 100 elements. I use `Ref` to access them conveniently. Otherwise, I could access `var1[50]`, or write a small accessor function.

yuyichao replied:
The correctness of the test doesn't and shouldn't depend on the cache layout so this is not necessary.

Also this doesn't ensure they are in different cache, since both of them are likely be allocated in neighboring cells in the pool. The right way to ensure they are in different cache is to add padding for non-atomic members, for atomic members, there's currently no way to do that.

eschnett replied:
I didn't spot that atomic types are reference types...

ArchRobison commented:
With respect to the question about `acq_rel` vs. `seq_cst` fences, we want the `seq_cst` flavor, because the `acq_rel` flavor [is a no-op on x86](https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html) and thus won't make Dekker-style protocols work.

This was surprising to me because a few years ago I had posed the question "how an I tell if mfence is sequentially consistent or not?" internally here at Intel, and we couldn't come up with a program that could exhibit the difference.  But we were thinking with our x86/Itanium hats on, which evidently brings in assumptions that don't hold for the C11/C++11 model.  Now I'm not clear on why a `std::atomic_thread_fence(std::memory_order_acq_rel)` wold be of any interest, except may to optimize a pair of acquire and release fences on some architectures.

eschnett commented:
@ArchRobison Thanks. Yes, I went with `seq_cst` fences.

yuyichao commented:
LGTM, need a squash.

eschnett committed:
Update atomic_rmw operations to use acq_rel ordering; introduce fence with seq_cst ordering

- set memory ordering for atomic loads to acquire
- set memory ordering for atomic stores to release
- set memory ordering for atomic rmw operations to acq_rel
- introduce fence instruction with seq_cst ordering
- add test cases

See #15106 for a related discussion.
